docs/discussion/Cosign Upgrade 2024.md

```diff
@@ -0,0 +1,468 @@
+
+Bob has a container image that is produced from a build pipeline and tested via a testing pipeline. After each pipeline, the image is signed with cosign using a SEPARATE key. Now the image has 2 cosign signatures. Bob trusts both keys and requires BOTH keys to be used. Bob wants to ensure all container images entering his K8s cluster are verified to have a valid cosign signature from his build AND test pipeline.
+
+- Bob installs Ratify on the cluster
+- Bob applies a new inline `KeyManagementSystem` CR onto the cluster, `inline-key-build` & `inline-key-test`
+```yaml
+apiVersion: config.ratify.deislabs.io/v1alpha1
+kind: KeyManagementSystem
+metadata:
+  name: inline-key-build
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  provider: inline
+  parameters:
+    type: key
+    value: |
+    ---------- BEGIN RSA KEY ------------
+    ******
+    ---------- END RSA KEY ------------
+    
+```
+```yaml
+apiVersion: config.ratify.deislabs.io/v1alpha1
+kind: KeyManagementSystem
+metadata:
+  name: inline-key-test
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  provider: inline
+  parameters:
+    type: key
+    value: |
+    ---------- BEGIN RSA KEY ------------
+    ******
+    ---------- END RSA KEY ------------
+    
+```
+
+- Bob applies the cosign verifier CR
+```yaml
+apiVersion: config.ratify.deislabs.io/v1beta1
+kind: Verifier
+metadata:
+  name: verifier-cosign
+  annotations:
+    helm.sh/hook: pre-install,pre-upgrade
+    helm.sh/hook-weight: "5"
+spec:
+  name: cosign
+  artifactTypes: application/vnd.dev.cosign.artifact.sig.v1+json
+  parameters:
+    keyManagementSystems:
+      - inline-key-build
+      - inline-key-test
+    keyEnforcement: all
+```
+- Bob attempts to deploy a pod from the vetted image that has 1 cosign signature, which is signed with key in KeyManagementSystem `inline-key-build`. Pod FAILS verification and blocked.
+- Bob attempts to deploy a pod from the vetted image that has 2 cosign signatures, which is signed with keys in KeyManagementSystem `inline-key-build` & `inline-key-test`. Pod passes verification and is created.
+
+> **NOTE**: Based on Cosign's own verification implementation, there does not seem to be a way to enforce that ALL signatures found are valid. This behavior is unchartered as far as we can tell and would require Ratify to determine if we want to support this directly. For that reason, the strict ALL signature scenario is not yet included.
+
+#### Extra Considerations
+- How does Kyverno handle multiple Keys?
+  - Each key specified is a separate cosign verification operation
+  - User specifies enforcement type by specifying `count` of verification operations that must succeed.
+    - e.g 2 keys specified and 2 signatures on the image. each verification is against both signatures for a single key
+    - If count is 2, then 2 verifications must return true
+- Why is `keyEnforcement` necessary?
```

Comment:
As per CC 1/24/24 discussion: lets leave `keyEnforcement` as a follow up task. Leave default multi key behavior to be "any".